lib/web_ui/lib/src/engine/bitmap_canvas.dart

```diff
@@ -664,6 +664,201 @@ class BitmapCanvas extends EngineCanvas with SaveStackTracking {
     picture.recordingCanvas.apply(this);
   }
 
+  // Vertex shader transforms pixel space [Vertices.positions] to
+  // final clipSpace -1..1 coordinates with inverted Y Axis.
+  static const _vertexShaderTriangle = '''
+      #version 300 es
+      layout (location=0) in vec4 position;
+      layout (location=1) in vec4 color;
+      uniform vec4 u_scale;
+      uniform vec4 u_shift;
+      
+      out vec4 vColor;
+      void main() {
+        gl_Position = (position * u_scale) + u_shift;
+        vColor = color;
```

Comment:
I think you should pull the color swizzing from the fragment shader into the vertex shader.

e.g. `vColor = color.zyxw;`

Then the fragment shader becomes pretty simple. This probably has a very minuscule effect on mobile GPUs today.

Reply:
thanks for suggestion. done.